### PR TITLE
Add support for guest numa node verification for hotplug

### PR DESCRIPTION
**Pull request by sathnaga**
Add method to validate cpus and memory in numa node during
hotplug operation

Signed-off-by: Satheesh Rajendran <sathnaga@linux.vnet.ibm.com>

### CODE REVIEW

**Comment by sathnaga** [comment]
Testing in progress...

**Comment by sathnaga** [comment]
Functionality failure.
```
avocado run cpuhotplug.positive_test.vcpu_set.live.vm_operate.no_operation --vt-type libvirt --vt-extra-params create_vm_libvirt=yes kill_vm_libvirt=yes env_cleanup=yes backup_image_before_testing=no restore_image_after_testing=no mem=8192 numa=yes guest_numa_nodes="node0 node1 node2 node3" numa_mem_node0=0 numa_cpus_node0=0,2 numa_nodeid_node0=0 numa_mem_node1=0 numa_cpus_node1=1,3 numa_nodeid_node1=1 numa_mem_node2=8388608 numa_cpus_node2=4-5 numa_nodeid_node2=2 numa_mem_node3=0 numa_cpus_node3=6-7 numa_nodeid_node3=3 smp=2 vcpu_threads=1 vcpu_cores=8 vcpu_sockets=1 vcpu_maxcpus=8 --vt-guest-os RHEL.7.devel.ppc64le
JOB ID     : 4d8b6a0469a93b59842e97a9e947841ebd13ab17
JOB LOG    : /home/results/job-2018-09-24T12.33-4d8b6a0/job.log
 (1/1) type_specific.powerkvm-libvirt.cpuhotplug.positive_test.vcpu_set.live.vm_operate.no_operation: FAIL: Test Failed (266.70 s)
RESULTS    : PASS 0 | ERROR 0 | FAIL 1 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB TIME   : 309.81 s
````

```
8-09-24 12:37:18,695 utils_misc       L2453 INFO | CPU quantity matched: 8
2018-09-24 12:37:18,697 utils_hotplug    L0539 DEBUG| Check guest numa
2018-09-24 12:37:18,698 virt_vm          L1146 DEBUG| Attempting to log into 'avocado-vt-vm1' (timeout 240s)
2018-09-24 12:37:18,701 virt_vm          L0862 DEBUG| Found/Verified IP 192.168.122.116 for VM avocado-vt-vm1 NIC 0
2018-09-24 12:37:21,217 client           L1126 DEBUG| Sending command: lscpu
2018-09-24 12:37:23,131 virt_vm          L1146 DEBUG| Attempting to log into 'avocado-vt-vm1' (timeout 240s)
2018-09-24 12:37:23,141 virt_vm          L0862 DEBUG| Found/Verified IP 192.168.122.116 for VM avocado-vt-vm1 NIC 0
2018-09-24 12:37:25,556 client           L1126 DEBUG| Sending command: count=0;[ -d /sys/devices/system/node/node0/ ] && cd /sys/devices/system/node/node0/;[ -d memory0 ] &&  for i in `ls -d memory*`;do [ -f $i/online ] && a=$(<$i/online) && count=$(( $count + $a ));a=0;done;echo $count
2018-09-24 12:37:25,675 client           L1126 DEBUG| Sending command: cat /sys/devices/system/memory/block_size_bytes
2018-09-24 12:37:27,338 utils_hotplug    L0568 ERROR| Mismatch in cpus in node 0: xml [0, 2] guest [0, 2, 6, 7]
2018-09-24 12:37:27,580 virt_vm          L1146 DEBUG| Attempting to log into 'avocado-vt-vm1' (timeout 240s)
2018-09-24 12:37:27,583 virt_vm          L0862 DEBUG| Found/Verified IP 192.168.122.116 for VM avocado-vt-vm1 NIC 0
2018-09-24 12:37:30,044 client           L1126 DEBUG| Sending command: count=0;[ -d /sys/devices/system/node/node1/ ] && cd /sys/devices/system/node/node1/;[ -d memory0 ] &&  for i in `ls -d memory*`;do [ -f $i/online ] && a=$(<$i/online) && count=$(( $count + $a ));a=0;done;echo $count
2018-09-24 12:37:30,188 client           L1126 DEBUG| Sending command: cat /sys/devices/system/memory/block_size_bytes
2018-09-24 12:37:31,298 virt_vm          L1146 DEBUG| Attempting to log into 'avocado-vt-vm1' (timeout 240s)
2018-09-24 12:37:31,302 virt_vm          L0862 DEBUG| Found/Verified IP 192.168.122.116 for VM avocado-vt-vm1 NIC 0
2018-09-24 12:37:34,446 client           L1126 DEBUG| Sending command: count=0;[ -d /sys/devices/system/node/node2/ ] && cd /sys/devices/system/node/node2/;[ -d memory0 ] &&  for i in `ls -d memory*`;do [ -f $i/online ] && a=$(<$i/online) && count=$(( $count + $a ));a=0;done;echo $count
2018-09-24 12:37:34,731 client           L1126 DEBUG| Sending command: cat /sys/devices/system/memory/block_size_bytes
2018-09-24 12:37:35,601 virt_vm          L1146 DEBUG| Attempting to log into 'avocado-vt-vm1' (timeout 240s)
2018-09-24 12:37:35,606 virt_vm          L0862 DEBUG| Found/Verified IP 192.168.122.116 for VM avocado-vt-vm1 NIC 0
2018-09-24 12:37:38,201 client           L1126 DEBUG| Sending command: count=0;[ -d /sys/devices/system/node/node3/ ] && cd /sys/devices/system/node/node3/;[ -d memory0 ] &&  for i in `ls -d memory*`;do [ -f $i/online ] && a=$(<$i/online) && count=$(( $count + $a ));a=0;done;echo $count
2018-09-24 12:37:38,323 client           L1126 DEBUG| Sending command: cat /sys/devices/system/memory/block_size_bytes
2018-09-24 12:37:39,204 utils_hotplug    L0568 ERROR| Mismatch in cpus in node 3: xml [6, 7] guest []
2018-09-24 12:37:39,207 utils_hotplug    L0578 ERROR| Mismatch in numa nodes expected nodes: 4 guest: 3
```

**Comment by sathnaga** [comment]
Reference: https://github.com/avocado-framework/avocado-vt/pull/1743

**Comment by sathnaga** [comment]
Changes for handling xml with no numa tags.
```
avocado run cpuhotplug.positive_test.vcpu_set.live.vm_operate.no_operation --vt-type libvirt --vt-extra-params create_vm_libvirt=yes kill_vm_libvirt=yes env_cleanup=yes backup_image_before_testing=no restore_image_after_testing=no mem=8192 smp=2 vcpu_threads=1 vcpu_cores=8 vcpu_sockets=1 vcpu_maxcpus=8 --vt-guest-os RHEL.7.devel.ppc64le
JOB ID     : 87aae84c383b626c460a0e6bf54e8bf5b64de969
JOB LOG    : /home/workspace/runAvocadoFVTTest/avocado-fvt-wrapper/results/job-2018-09-24T14.12-87aae84/job.log
 (1/1) type_specific.powerkvm-libvirt.cpuhotplug.positive_test.vcpu_set.live.vm_operate.no_operation: PASS (224.35 s)
```

**Comment by sathnaga** [comment]
Result with no numa tags in guest_xml:
```
avocado run cpuhotplug.positive_test.vcpu_set.live.vm_operate.no_operation --vt-type libvirt --vt-extra-params create_vm_libvirt=yes kill_vm_libvirt=yes env_cleanup=yes backup_image_before_testing=no restore_image_after_testing=no mem=8192 smp=2 vcpu_threads=1 vcpu_cores=8 vcpu_sockets=1 vcpu_maxcpus=8 --vt-guest-os RHEL.7.devel.ppc64le
JOB ID     : 87aae84c383b626c460a0e6bf54e8bf5b64de969
JOB LOG    : /home//results/job-2018-09-24T14.12-87aae84/job.log
 (1/1) type_specific.powerkvm-libvirt.cpuhotplug.positive_test.vcpu_set.live.vm_operate.no_operation: PASS (224.35 s)
```

Result with numa tags in guest_xml:

```

# avocado run cpuhotplug.positive_test.vcpu_set.live.vm_operate.no_operation --vt-type libvirt --vt-extra-params create_vm_libvirt=yes kill_vm_libvirt=yes env_cleanup=yes backup_image_before_testing=no restore_image_after_testing=no mem=8192 numa=yes guest_numa_nodes="node0 node1 node2 node3" numa_mem_node0=0 numa_cpus_node0=0,2 numa_nodeid_node0=0 numa_mem_node1=0 numa_cpus_node1=1,3 numa_nodeid_node1=1 numa_mem_node2=8388608 numa_cpus_node2=4-5 numa_nodeid_node2=2 numa_mem_node3=0 numa_cpus_node3=6-7 numa_nodeid_node3=3 smp=2 vcpu_threads=1 vcpu_cores=8 vcpu_sockets=1 vcpu_maxcpus=8 --vt-guest-os RHEL.7.devel.ppc64le
JOB ID     : 802e0d7ec92d0c471110844663398b68b2cd64f3
JOB LOG    : /home//results/job-2018-09-24T14.25-802e0d7/job.log
 (1/1) type_specific.powerkvm-libvirt.cpuhotplug.positive_test.vcpu_set.live.vm_operate.no_operation: FAIL: Test Failed (318.25 s)
RESULTS    : PASS 0 | ERROR 0 | FAIL 1 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB TIME   : 331.61 s
```

File: virttest/utils_hotplug.py [hunk 1]
```diff
@@ -461,10 +461,13 @@ def check_vcpu_value(vm, exp_vcpu, vcpupin=None, option="", guest_agent=False):
     if not check_xmlcount(vm, exp_vcpu, option):
         final_result = False
 
-    # 1.5 Check inside the guest
     if vm.is_alive() and (not vm.is_paused()) and "live" in option:
+        # 1.5 Check inside the guest
         if not utils_misc.check_if_vm_vcpu_match(exp_vcpu['guest_live'], vm):
             final_result = False
+        #1.6 Check guest numa
```
Comment (balamuruhans) [inline virttest/utils_hotplug.py]: a space after #

**Pull request merged**